File: bridge/internal/configdir/configdir_test.go
package configdir_test

import (
	"testing"

	"github.com/Vex-Foundation/vex/bridge/internal/configdir"
	"github.com/Vex-Foundation/vex/bridge/internal/vectors"
)

// The config-directory resolver is the highest-divergence seam on this wire:
// the endpoint hash is taken over its result, so a difference of one separator
// between this package and the two Node owners is a bridge that dials a path
// the app never bound. The three run the SAME table.
func TestResolveMatchesGoldenVectors(t *testing.T) {
	file, err := vectors.Load()
	if err != nil {
		t.Fatalf("loading the golden vectors: %v", err)
	}
	if len(file.ConfigDir.Cases) == 0 {
		t.Fatal("the fixture carries no configDir cases; the resolver would be unpinned")
	}
	if file.ConfigDir.AppName != configdir.AppName {
		t.Fatalf("app name: fixture %q, package %q", file.ConfigDir.AppName, configdir.AppName)
	}

	for _, testCase := range file.ConfigDir.Cases {
		t.Run(testCase.Name, func(t *testing.T) {
			got := configdir.Resolve(
				vectors.GOOS(testCase.Platform),
				testCase.Homedir,
				configdir.Env(testCase.Env),
			)
			if got != testCase.Expect {
				t.Fatalf("resolved %q, the contract says %q", got, testCase.Expect)
			}
		})
	}
}

// The specific defect the hardening closed: an empty XDG_CONFIG_HOME used to
// join into the RELATIVE path "vex", putting the config directory in whatever
// working directory the launcher happened to have.
func TestEmptyEnvNeverProducesARelativeDirectory(t *testing.T) {
	for name, env := range map[string]configdir.Env{
		"empty XDG_CONFIG_HOME":  {"XDG_CONFIG_HOME": ""},
		"empty VEX_CONFIG_DIR":   {"VEX_CONFIG_DIR": ""},
		"blank-ish XDG relative": {"XDG_CONFIG_HOME": "config"},
	} {
		t.Run(name, func(t *testing.T) {
			got := configdir.Resolve("linux", "/home/alice", env)
			if got != "/home/alice/.config/vex" {
				t.Fatalf("got %q, want the home fallback", got)
			}
		})
	}
	got := configdir.Resolve("windows", `C:\Users\alice`, configdir.Env{"APPDATA": ""})
	if got != `C:\Users\alice\AppData\Roaming\vex` {
		t.Fatalf("windows empty APPDATA gave %q", got)
	}
}

// An accepted VEX_CONFIG_DIR is returned VERBATIM. A trailing separator is a
// different string, therefore a different hash, therefore a different endpoint,
// and the resolver must not tidy it away on one side of the wire only.
func TestAcceptedOverrideIsVerbatim(t *testing.T) {
	for _, value := range []string{"/srv/vexstate", "/srv/vexstate/", "/srv//vexstate/./"} {
		got := configdir.Resolve("linux", "/home/alice", configdir.Env{"VEX_CONFIG_DIR": value})
		if got != value {
			t.Fatalf("override %q came back as %q", value, got)
		}
	}
}

File: vex-app/src/main/paths/__tests__/config-dir.test.ts
/**
 * The CONFIG_DIR resolver, executed against the GOLDEN VECTORS.
 *
 * This is not a mirror-check between two TypeScript files any more. The Vex
 * Studio endpoint discriminator is a SHA-256 over this directory, and three
 * independent implementations derive it: this one, `src/config/paths.ts`, and
 * the standalone Go bridge (`bridge/internal/configdir`). A drift of one
 * separator between any two of them is a bridge that dials a path the app
 * never bound, so all three run the same table from
 * `studio-mcp/bridge-endpoint-vectors.json`.
 */

import { readFileSync } from "node:fs";
import path from "node:path";

import { describe, expect, it } from "vitest";
import {
  ELECTRON_STATE_DIR,
  ENV_FILE,
  INSTALL_ID_FILE,
  PG_PASSWORD_FILE,
  SECRETS_DIR,
  SETUP_COMPLETE_FILE,
  resolveConfigDir,
} from "../config-dir.js";

interface ConfigDirCase {
  readonly name: string;
  readonly platform: NodeJS.Platform;
  readonly homedir: string;
  readonly env: Record<string, string>;
  readonly expect: string;
}

interface Vectors {
  readonly configDir: {
    readonly appName: string;
    readonly cases: readonly ConfigDirCase[];
  };
}

const VECTORS_PATH = path.resolve(
  __dirname,
  "..", "..", "..", "..", "..",
  "src", "vex-agent", "tools", "tool-surface-spec", "studio-mcp",
  "bridge-endpoint-vectors.json",
);

const vectors = JSON.parse(readFileSync(VECTORS_PATH, "utf8")) as Vectors;

describe("resolveConfigDir golden vectors", () => {
  it("the fixture carries a resolver matrix at all", () => {
    expect(vectors.configDir.cases.length).toBeGreaterThan(15);
    expect(vectors.configDir.appName).toBe("vex");
  });

  it.each(vectors.configDir.cases)("$name", (testCase) => {
    expect(
      resolveConfigDir({
        platform: testCase.platform,
        homedir: testCase.homedir,
        env: testCase.env,
      }),
    ).toBe(testCase.expect);
  });

  it("every documented environment name appears in the matrix", () => {
    const covered = new Set<string>();
    for (const testCase of vectors.configDir.cases) {
      for (const name of Object.keys(testCase.env)) covered.add(name);
    }
    expect([...covered].sort()).toEqual(["APPDATA", "VEX_CONFIG_DIR", "XDG_CONFIG_HOME"]);
  });
});

describe("the hardening: empty and relative directory variables are UNSET", () => {
  // The defect this closed: `env["XDG_CONFIG_HOME"] ?? join(home, ".config")`
  // accepted an empty string, so `join("", "vex")` produced the RELATIVE path
  // "vex" and the config directory landed in whatever cwd the launcher had.
  // The XDG Base Directory specification requires an empty value to be ignored.
  it.each([
    ["XDG_CONFIG_HOME", ""],
    ["XDG_CONFIG_HOME", "relative/config"],
    ["XDG_CONFIG_HOME", "."],
  ])("linux %s=%j falls back to the home directory", (name, value) => {
    const resolved = resolveConfigDir({
      platform: "linux",
      homedir: "/home/kuba",
      env: { [name]: value },
    });
    expect(resolved).toBe("/home/kuba/.config/vex");
    expect(path.posix.isAbsolute(resolved)).toBe(true);
  });

  it.each([
    ["APPDATA", ""],
    ["APPDATA", "AppData\\Roaming"],
  ])("win32 %s=%j falls back to the home directory", (name, value) => {
    expect(
      resolveConfigDir({
        platform: "win32",
        homedir: "C:\\Users\\kuba",
        env: { [name]: value },
      }),
    ).toBe("C:\\Users\\kuba\\AppData\\Roaming\\vex");
  });

  it("never returns a relative path for any empty-variable combination", () => {
    for (const platform of ["linux", "darwin", "win32"] as const) {
      const home = platform === "win32" ? "C:\\Users\\kuba" : "/home/kuba";
      const resolved = resolveConfigDir({
        platform,
        homedir: home,
        env: { VEX_CONFIG_DIR: "", XDG_CONFIG_HOME: "", APPDATA: "" },
      });
      const target = platform === "win32" ? path.win32 : path.posix;
      expect(target.isAbsolute(resolved)).toBe(true);
      expect(resolved).not.toBe("vex");
    }
  });
});

describe("the override contract", () => {
  it("honours an absolute VEX_CONFIG_DIR (test/CI escape hatch)", () => {
    expect(
      resolveConfigDir({
        platform: "linux",
        homedir: "/home/kuba",
        env: { VEX_CONFIG_DIR: "/tmp/vex-e2e-abc123" },
      }),
    ).toBe("/tmp/vex-e2e-abc123");
  });

  it("returns an accepted override VERBATIM, trailing separator included", () => {
    // A trailing separator is a different string, therefore a different hash,
    // therefore a different endpoint. Tidying it on one side of the wire only
    // is the exact drift the contract forbids.
    for (const value of ["/srv/vexstate", "/srv/vexstate/", "/srv//vexstate/./"]) {
      expect(
        resolveConfigDir({ platform: "linux", homedir: "/home/kuba", env: { VEX_CONFIG_DIR: value } }),
      ).toBe(value);
    }
  });

  it("uses the same lowercase `vex` app name as the engine resolver", () => {
    // src/config/paths.ts declares APP_NAME = "vex". A capitalised variant here
    // would silently split user state across two directories.
    const dir = resolveConfigDir({ platform: "linux", homedir: "/home/x", env: {} });
    expect(dir.endsWith("/vex")).toBe(true);
    expect(dir.endsWith("/Vex")).toBe(false);
  });
});

describe("derived path constants", () => {
  it("places the Electron-private state nested under CONFIG_DIR", () => {
    expect(ELECTRON_STATE_DIR.endsWith(path.join("vex", ".electron-state"))).toBe(true);
  });

  it("places shared resources at CONFIG_DIR root, not under .electron-state", () => {
    expect(ENV_FILE.includes(".electron-state")).toBe(false);
    expect(INSTALL_ID_FILE.includes(".electron-state")).toBe(false);
    expect(SETUP_COMPLETE_FILE.includes(".electron-state")).toBe(false);
    expect(SECRETS_DIR.includes(".electron-state")).toBe(false);
    expect(PG_PASSWORD_FILE.includes(".electron-state")).toBe(false);
  });

  it("PG password lives at CONFIG_DIR/local-infra/secrets/pg_password", () => {
    expect(PG_PASSWORD_FILE.endsWith(path.join("local-infra", "secrets", "pg_password"))).toBe(true);
  });
});
